File: algorithms/min_stack.py
#!/usr/bin/env python
# encoding: utf8

# Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
#     push(x) -- Push element x onto stack.
#     pop() -- Removes the element on top of the stack.
#     top() -- Get the top element.
#     getMin() -- Retrieve the minimum element in the stack.

__author__ = "Zhu Xianfeng <xianfeng.zhu@gmail.com>"

class MinStack:
    def __init__(self):
        self.stack = []
        self.mins  = []

    # @param x, an integer
    # @return an integer
    def push(self, x):
        self.stack.append(x)
        if len(self.mins) == 0:
            self.mins.append(x)
        elif self.mins[-1] >= x:
            self.mins.append(x)
        else:
            # minx = self.mins[-1]
            # self.mins.append(minx)
            pass
        return x

    # @return nothing
    def pop(self):
        if self.stack[-1] == self.mins[-1]:
            del self.mins[-1:]
        del self.stack[-1:]

    # @return an integer
    def top(self):
        return self.stack[-1]

    # @return an integer
    def getMin(self):
        return self.mins[-1]

def main():
    stack = MinStack()
    for i in range(10000):
        stack.push(i)
    stack.pop()
    print stack.getMin()


if __name__ == "__main__":
    main()

File: algorithms/min_stack.cpp
// 155. Min Stack: https://leetcode.com/problems/min-stack
// Author: xianfeng.zhu@gmail.com

#include <algorithm>
#include <climits>

#include "gmock/gmock.h"
#include "gtest/gtest.h"

class MinStack {
  struct Node {
    int val;
    int min;
    Node* next;
    Node(int _val, int _min, Node* _next = nullptr) : val(_val), min(_min), next(_next) {}
  };

 public:
  MinStack() : head_(nullptr) {}

  void push(int x) {
    if (head_ == nullptr) {
      head_ = new Node(x, x);
    } else {
      head_ = new Node(x, std::min(head_->min, x), head_);
    }
  }

  void pop() {
    auto* ptr = head_;
    head_ = head_->next;
    delete ptr;
  }

  int top() { return head_->val; }

  int getMin() { return head_->min; }

 private:
  Node* head_;
};

TEST(Solution, MinStack) {
  MinStack min_stack;
  min_stack.push(-2);
  min_stack.push(0);
  min_stack.push(-3);
  EXPECT_EQ(min_stack.getMin(), -3);

  min_stack.pop();
  EXPECT_EQ(min_stack.top(), 0);
  EXPECT_EQ(min_stack.getMin(), -2);
}
